File: _tools/encrypt-batch.rb
#!/usr/bin/env ruby
# encrypt-batch.rb — Ruby port of encrypt-batch.js
# Uses PBKDF2-SHA256 + AES-256-GCM matching the JS implementation exactly.
# Blob layout: salt[16] + iv[12] + ciphertext (same as encrypt-batch.js)
#
# Usage: ruby _tools/encrypt-batch.rb <passphrase> <directory>

require 'openssl'
require 'base64'
require 'fileutils'

passphrase = ARGV[0]
target_dir = ARGV[1]

if passphrase.nil? || target_dir.nil?
  warn 'Usage: ruby _tools/encrypt-batch.rb <passphrase> <directory>'
  exit 1
end

unless Dir.exist?(target_dir)
  warn "Directory not found: #{target_dir}"
  exit 1
end

# ── Crypto ──────────────────────────────────────────────────────────────────

def encrypt_content(passphrase, plaintext)
  salt = OpenSSL::Random.random_bytes(16)
  iv   = OpenSSL::Random.random_bytes(12)

  key = OpenSSL::PKCS5.pbkdf2_hmac(
    passphrase,
    salt,
    100_000,
    32,          # 256-bit key
    'SHA256'
  )

  cipher = OpenSSL::Cipher.new('aes-256-gcm')
  cipher.encrypt
  cipher.key = key
  cipher.iv  = iv

  ciphertext = cipher.update(plaintext.encode('UTF-8')) + cipher.final
  tag        = cipher.auth_tag   # 16-byte GCM authentication tag (matches Web Crypto)

  blob = salt + iv + ciphertext + tag
  Base64.strict_encode64(blob)
end

# ── Front-matter helpers ─────────────────────────────────────────────────────

def split_front_matter(content)
  m = content.match(/\A---\r?\n(.*?)\r?\n---\r?\n?(.*)\z/m)
  return [{}, content] unless m

  fields = {}
  m[1].each_line do |line|
    kv = line.match(/^([a-zA-Z_][\w-]*)\s*:\s*(.*)$/)
    next unless kv
    fields[kv[1]] = kv[2].strip.gsub(/\A["']|["']\z/, '')
  end
  [fields, m[2]]
end

NEEDS_QUOTES_RE = /[:#"']/

def needs_quotes?(val)
  val.is_a?(String) && (
    val.match?(NEEDS_QUOTES_RE) ||
    val.start_with?(' ') ||
    val.end_with?(' ') ||
    val.empty? ||
    val.length > 200
  )
end

def build_front_matter(fields)
  lines = ['---']
  fields.each do |k, v|
    next if v.nil?
    s = v.to_s
    lines << (needs_quotes?(s) ? "#{k}: \"#{s.gsub('"', '\\"')}\"" : "#{k}: #{s}")
  end
  lines << '---'
  lines.join("\n") + "\n"
end

# ── Markdown cleaner ─────────────────────────────────────────────────────────

def clean_markdown(md)
  md
    .gsub(/^## Table of contents[\s\S]*?^---$/m, '---')  # remove TOC section
    .gsub(/^\{:.*?\}\s*$/m, '')                           # remove Kramdown attribute lines
    .gsub(/\{%[^%]*%\}/, '')                              # remove Liquid tags
    .gsub(/\n{3,}/, "\n\n")                               # collapse blank lines
    .strip
end

# ── Process files ────────────────────────────────────────────────────────────

files = Dir.glob(File.join(target_dir, '*.md')).sort

if files.empty?
  warn "No .md files found in #{target_dir}"
  exit 1
end

puts "\n🔐  Encrypting #{files.length} files in #{File.basename(target_dir)}/\n\n"

files.each do |file_path|
  raw = File.read(file_path, encoding: 'UTF-8')
  fields, body = split_front_matter(raw)

  clean_body = clean_markdown(body)
  if clean_body.empty?
    puts "  ⚠  Skipping #{File.basename(file_path)} — empty body"
    next
  end

  blob = encrypt_content(passphrase, clean_body)

  new_fields = {}
  new_fields['layout']       = 'protected'
  new_fields['title']        = fields['title']       if fields['title']
  new_fields['parent']       = fields['parent']      if fields['parent']
  new_fields['nav_order']    = fields['nav_order']   if fields['nav_order']
  new_fields['has_children'] = fields['has_children'] if fields['has_children']
  new_fields['permalink']    = fields['permalink']   if fields['permalink']
  new_fields['encrypted']    = blob

  File.write(file_path, build_front_matter(new_fields), encoding: 'UTF-8')
  puts "  ✓  #{File.basename(file_path)}"
end

puts "\n✅  Done — #{files.length} files encrypted.\n"

File: _tools/encrypt-batch.py
#!/usr/bin/env python3
"""
encrypt-batch.py — Python port of encrypt-batch.js
Blob layout: salt[16] + iv[12] + ciphertext+tag (same as Web Crypto AES-GCM)
PBKDF2-HMAC-SHA256, 100000 iterations, 32-byte key.

Usage: python3 _tools/encrypt-batch.py <passphrase> <directory>
"""

import sys, os, re, glob, base64, hashlib, secrets

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

def derive_key(passphrase: str, salt: bytes) -> bytes:
    return hashlib.pbkdf2_hmac('sha256', passphrase.encode('utf-8'), salt, 100_000, dklen=32)

def encrypt_content(passphrase: str, plaintext: str) -> str:
    salt = secrets.token_bytes(16)
    iv   = secrets.token_bytes(12)
    key  = derive_key(passphrase, salt)
    # AESGCM.encrypt returns ciphertext + 16-byte tag (matches Web Crypto)
    ct_and_tag = AESGCM(key).encrypt(iv, plaintext.encode('utf-8'), None)
    blob = salt + iv + ct_and_tag
    return base64.b64encode(blob).decode('ascii')

# ── Front-matter helpers ─────────────────────────────────────────────────────

def split_front_matter(content: str):
    m = re.match(r'^---\r?\n(.*?)\r?\n---\r?\n?(.*)', content, re.DOTALL)
    if not m:
        return {}, content
    fields = {}
    for line in m.group(1).splitlines():
        kv = re.match(r'^([a-zA-Z_][\w-]*)\s*:\s*(.*)$', line)
        if kv:
            fields[kv.group(1)] = kv.group(2).strip().strip('"\'')
    return fields, m.group(2)

def needs_quotes(val: str) -> bool:
    return bool(re.search(r'[:#"\']', val)) or \
           val.startswith(' ') or val.endswith(' ') or \
           val == '' or len(val) > 200

def build_front_matter(fields: dict) -> str:
    lines = ['---']
    for k, v in fields.items():
        if v is None:
            continue
        s = str(v)
        if needs_quotes(s):
            lines.append(f'{k}: "{s.replace(chr(34), chr(92)+chr(34))}"')
        else:
            lines.append(f'{k}: {s}')
    lines.append('---')
    return '\n'.join(lines) + '\n'

# ── Markdown cleaner ─────────────────────────────────────────────────────────

def clean_markdown(md: str) -> str:
    md = re.sub(r'^## Table of contents[\s\S]*?^---$', '---', md, flags=re.MULTILINE)
    md = re.sub(r'^\{:.*?\}\s*$', '', md, flags=re.MULTILINE)
    md = re.sub(r'\{%[^%]*%\}', '', md)
    md = re.sub(r'\n{3,}', '\n\n', md)
    return md.strip()

# ── Main ─────────────────────────────────────────────────────────────────────

if len(sys.argv) < 3:
    print('Usage: python3 _tools/encrypt-batch.py <passphrase> <directory>', file=sys.stderr)
    sys.exit(1)

passphrase = sys.argv[1]
target_dir = sys.argv[2]

if not os.path.isdir(target_dir):
    print(f'Directory not found: {target_dir}', file=sys.stderr)
    sys.exit(1)

files = sorted(glob.glob(os.path.join(target_dir, '*.md')))
if not files:
    print(f'No .md files found in {target_dir}', file=sys.stderr)
    sys.exit(1)

print(f'\n🔐  Encrypting {len(files)} files in {os.path.basename(target_dir)}/\n')

for file_path in files:
    with open(file_path, 'r', encoding='utf-8') as f:
        raw = f.read()

    fields, body = split_front_matter(raw)
    clean_body = clean_markdown(body)

    if not clean_body:
        print(f'  ⚠  Skipping {os.path.basename(file_path)} — empty body')
        continue

    blob = encrypt_content(passphrase, clean_body)

    new_fields = {'layout': 'protected'}
    for key in ('title', 'parent', 'nav_order', 'has_children', 'permalink'):
        if fields.get(key):
            new_fields[key] = fields[key]
    new_fields['encrypted'] = blob

    with open(file_path, 'w', encoding='utf-8') as f:
        f.write(build_front_matter(new_fields))

    print(f'  ✓  {os.path.basename(file_path)}')

print(f'\n✅  Done — {len(files)} files encrypted.\n')
